b2b/keycloak_admin_api.py

```diff
@@ -0,0 +1,427 @@
+        """
+
+        self.base_url = settings.KEYCLOAK_BASE_URL
+        if not self.base_url:
+            msg = "KEYCLOAK_BASE_URL setting is not configured."
+            raise KeycloakAdminImproperlyConfiguredError(msg)
+        self.base_url = urljoin(self.base_url, "/admin/realms/")
+        self._realm = settings.KEYCLOAK_REALM_NAME
+        if not self._realm:
+            msg = "KEYCLOAK_REALM_NAME setting is not configured."
+            raise KeycloakAdminImproperlyConfiguredError(msg)
+
+        realm_discovery = settings.KEYCLOAK_DISCOVERY_URL
+        if not realm_discovery:
+            msg = "KEYCLOAK_DISCOVERY_URL setting is not configured."
+            raise KeycloakAdminImproperlyConfiguredError(msg)
+        client_id = settings.KEYCLOAK_ADMIN_CLIENT_ID
+        if not client_id:
+            msg = "KEYCLOAK_ADMIN_CLIENT_ID setting is not configured."
+            raise KeycloakAdminImproperlyConfiguredError(msg)
+        client_secret = settings.KEYCLOAK_ADMIN_CLIENT_SECRET
+        if not client_secret:
+            msg = "KEYCLOAK_ADMIN_CLIENT_SECRET setting is not configured."
+            raise KeycloakAdminImproperlyConfiguredError(msg)
+
+        self.skip_verify = settings.KEYCLOAK_ADMIN_CLIENT_NO_VERIFY_SSL or False
+
+        openid_configuration = requests.get(
+            realm_discovery,
+            timeout=60,
+            verify=not self.skip_verify,
+        )
+        openid_configuration.raise_for_status()
+        self.openid_configuration = openid_configuration.json()
+
+        self.oauth_session = OAuth2Session(
+            client_id=client_id,
+            client_secret=client_secret,
+            token_endpoint=self.openid_configuration["token_endpoint"],
+            scope=settings.KEYCLOAK_ADMIN_CLIENT_SCOPES,
+            verify=not self.skip_verify,
+        )
+        self.token = self.oauth_session.fetch_token(
+            self.openid_configuration["token_endpoint"],
+            grant_type="client_credentials",
+        )
+
+    def realmify_url(self, url_path):
+        """
+        Prefix the given URL path with the realm base path.
+
+        The admin API URLs are all /admin/realms/ and then _usually_ a realm
+        name.
+
+        Args:
+        - url_path: The URL path to prefix with the realm base path.
+        Returns:
+        - The full URL path with the realm base path prefixed.
+        """
+
+        return urljoin(self.base_url, f"{self._realm}/{url_path}")
+
+    def request(self, method, url_path, *, skip_realmify=False, **kwargs):
+        """Perform an HTTP request against the given URL path."""
+
+        request_url = (
+            urljoin(self.base_url, url_path)
+            if skip_realmify
+            else self.realmify_url(url_path)
+        )
+
+        return (
+            self.oauth_session.request(method, request_url, **kwargs)
+            if self.oauth_session
+            else None
+        )
```

Comment:
i think this might be worth splitting this into a `request`. `realm_request`, and `_request` (for the common stuff) so the behavior is split at the function level rather than needing to pass an arg, I think that documents the intent better.